File: solutions/kotlin/react/3/src/main/kotlin/React.kt
import java.util.IdentityHashMap

class Reactor<T>() {
    private val dag = mutableListOf<Cell<T>>()

    private fun invokeChain(id: Int) {
        dag.asSequence().drop(id + 1).forEach{ cell ->
            cell.value
        }
    }
    
    fun InputCell(initial: T): InputCell<T> {
        val id = dag.size
        val input = InputCell(id = id, initial = initial) {
            invokeChain(id)
        }
        dag.add(input)
        return input
    }

    fun ComputeCell(vararg cells: Cell<T>, compute: (List<T>) -> T): ComputeCell<T> {
        val id = dag.size
        val com = ComputeCell(id = id) {
            compute(cells.map(Cell<T>::value))
        }
        dag.add(com)
        return com
    }
}

interface Subscription {
    fun cancel()
}

interface Cell<T> {
    val id: Int
    val value: T
}

class InputCell<T>(
    override val id: Int, initial: T, val runner: () -> Unit
): Cell<T> {
    override var value: T = initial
        set(newValue) {
            if (field == newValue) {
                return
            }
            field = newValue
            runner()
        }
}

class ComputeCell<T>(override val id: Int, val compute: () -> T): Cell<T> {
    private val registry = IdentityHashMap<Canceler<T>, (T) -> Unit>()
    override var value: T = compute()
        get() {
            val newValue = compute()
            if (field == newValue) {
                return field
            }
            field = newValue
            registry.forEach{ (_, callback) ->
                callback(field)
            }
            return field
        }

    fun addCallback(callback: (T) -> Unit): Canceler<T> {
        val newCanceler = Canceler(callback, registry)
        registry.put(newCanceler, callback)
        return newCanceler
    }
}

data class Canceler<T>(
    val callback: (T) -> Unit, val registry: MutableMap<Canceler<T>, (T) -> Unit>
): Subscription {
    override fun cancel() {
        registry.remove(this)
    }
}

File: solutions/go/react/21/react.go
package react

type reactor struct {
    cells		[]Cell
    runner		func(int)
}

type inputCell struct {
    id			int
    value		int
    runner		func(int)
}

type computeCell struct {
    value		int
    compute		func() int
    registry	map[Canceler]func(int)
}

type canceler struct {
    cancel		func()
}

func New() Reactor {
    r := &reactor{
        cells:	make([]Cell, 0, 128),
    }
    r.runner = func(id int) {
        if id == len(r.cells) - 1 {
            return
        }
        for _, cell := range r.cells[id + 1:] {
            _ = cell.Value()
        }
    }
    return r
}

func (r *reactor) CreateInput(initial int) InputCell {
    input := &inputCell{
        id:		len(r.cells),
        value:	initial,
        runner:	r.runner,
    }
    r.cells = append(r.cells, input)
    return input
}

func (r *reactor) CreateCompute1(dep Cell, compute func(int) int) ComputeCell {
    com := &computeCell{
        value:		compute(dep.Value()),
        compute:	func() int { return compute(dep.Value()) },
        registry:	make(map[Canceler]func(int)),
    }
    r.cells = append(r.cells, com)
    return com
}

func (r *reactor) CreateCompute2(dep1, dep2 Cell, compute func(int, int) int) ComputeCell {
    com := &computeCell{
        value:		compute(dep1.Value(), dep2.Value()),
        compute:	func() int { return compute(dep1.Value(), dep2.Value()) },
        registry:	make(map[Canceler]func(int)),
    }
    r.cells = append(r.cells, com)
    return com
}

func (c *inputCell) Value() int {
    return c.value
}

func (c *inputCell) SetValue(value int) {
    if c.value == value {
        return
    }
    c.value = value
    c.runner(c.id)
}

func (c *computeCell) Value() int {
    newValue := c.compute()
    if c.value == newValue {
        return c.value
    }
    c.value = newValue
    for _, callback := range c.registry {
        callback(c.value)
    }
    return c.value
}

func (c *computeCell) AddCallback(callback func(int)) Canceler {
    newCanceler := &canceler{}
    newCanceler.cancel = func() {
        delete(c.registry, newCanceler)
    }
    c.registry[newCanceler] = callback
    return newCanceler
}

func (c *canceler) Cancel() {
    c.cancel()
}